src/v/storage/tests/utils/log_gap_analysis.h

```diff
@@ -0,0 +1,49 @@
+/*
+ * Copyright 2022 Redpanda Data, Inc.
+ *
+ * Use of this software is governed by the Business Source License
+ * included in the file licenses/BSL.md
+ *
+ * As of the Change Date specified in that file, in accordance with
+ * the Business Source License, use of this software will be governed
+ * by the Apache License, Version 2.0
+ */
+#pragma once
+
+#include "model/fundamental.h"
+#include "model/record_batch_reader.h"
+
+namespace storage {
+
+// Utility to analyze offset gaps in logs, for testing compaction logic.
+
+/**
+ * Results of analysis of message offset gaps within a log instance.
+ * This struct is used to return the total `num_gaps` detected, as well as the
+ * minimum and maximum offsets that were missing.
+ * This is useful for observing which offsets were affected by compaction.
+ */
+struct log_gap_analysis {
```

Comment:
nit: another useful thing would be the analysis of records read and unique keys read. For example, we produce 10000 records with 200 unique keys and consume 200 records with 200 unique keys. So we see how much we compact data.